# Metabolic control of progenitor cell propagation during _Drosophila_ tracheal remodeling

## RNA-seq analysis scripts

### Fig. 2 and Fig. S1

This is the repository for scripts used in paper of

*Metabolic control of progenitor cell propagation during *Drosophila* tracheal remodeling*

## GO

1. Use R package `DESeq2` to perform differential expression analysis.(see `heatmap of all genes`)
2. Generate a differentially expressed genes(DEGs) list.
3. Perform GO analysis for DEGs list and fetch pathway statistical information with geneRatio and pvalue from `DAVID Bioinformatics Resources 6.8`.
4. Draw a bubble chart to display our GO result.

### Input

A .csv table fetched from `DAVID Bioinformatics Resources 6.8` with column names as below:

```
pathway,Count,PValue,GeneRatio
```

**example**

```
DAVID using FlybaseID with generatio.csv
```

### Usage

run this script via R-studio and R:

```
ggplot.R
```

### Output

A .pdf format plot.

## Heatmap of all genes

1. Use R package `DESeq2` to perform differential expression analysis.
2. Generate a dds matrix.
3. Analyse with DEseq.
4. Export result files and make heatmaps for all genes.

### Input

A raw counts matrix generated by upstream quantitative analysis tool `featureCounts`is required.
**example**

```
L3vs0hr_Geneid for heatmap.csv
```

### Usage

run this script via R-studio and R:

```
heatmap_whole.R
```

### Output

A result file contains differential expression foldchange and adjusted pvalue for all genes.
A heatmap indicates the differential expression foldchange of all genes.

## Heatmap of specific genes

1. Use R package `DESeq2` to perform differential expression analysis.
2. Generate a dds matrix.
3. Analyse with DEseq.
4. Specify some candidate genes, then export result files and make heatmaps for these genes.

### Input

A raw counts matrix generated by upstream quantitative analysis tool `featureCounts`is required.
**example**

```
L3vs0hr_Geneid for heatmap.csv
```

### Usage

run this script via R-studio and R:

```
heatmap_specific_genes.R
```

### Output

A result file contains differential expression foldchange and adjusted pvalue for all genes.
A heatmap indicates differential expression foldchange of some candidate genes.

## Volcano plot

1. Use R package `DESeq2` to perform differential expression analysis.
2. Generate a dds matrix.
3. Analyse with DEseq.
4. Draw a volcano plot that indicates some of the most significant differentially expressed genes.

### Input

A `DeSeq2` result file with differential expression foldchange and adjusted pvalue for all genes is required.
**example**

```
0hrvsL3 for volcano.csv
```

### Usage

run this script via R-studio and R:

```
Volcano plot.R
```

### Output

A volcano plot that indicates some of the most significant differentially expressed genes.

### File tree

```
.
│  README.html
│  README.md
│
├─GO
│      DAVID using FlybaseID with generatio.csv
│      ggplot.R
│
├─Heatmap of all genes
│      heatmap_whole.R
│      L3vs0hr_Geneid for heatmap.csv
│
├─Heatmap of specific gene
│      heatmap_specific_genes.R
│      L3vs0hr_Geneid for heatmap.csv
│
└─Volcano plot
        0hrvsL3 for volcano.csv
        Volcano plot.R
```

> Huang DW, Sherman BT, Lempicki RA. Systematic and integrative analysis of large gene lists using DAVID Bioinformatics Resources. Nature Protoc. 2009;4(1):44-57.
