## ChIP-seq scripts

### HOMER

HOMER (Hypergeometric Optimization of Motif EnRichment) is a suite of tools for Motif Discovery and next-gen sequencing analysis. It is a collection of command line programs for **UNIX-style** operating systems written in Perl and C++. HOMER was primarily written as a de novo motif discovery algorithm and is well suited for finding 8-20 bp motifs in large scale genomics data.

We performed HOMER analysis on Ubuntu 18.04.2 LTS with 8 cores CPU(11th Gen Intel i7-1165G7 (8) @ 2.80GHz) and 32GB RAM.

#### Input

A .bed file generated by MACS2 after peak calling is required.

**example**

```
sp1.bed
```

#### Usage: Bash command

```
# After adding findMotifsGenome.pl to $PATH
findMotifsGenome.pl <peak/BED file> <genome> <output directory> -size # [options]
```

**example**

```
findMotifsGenome.pl sp1.bed dm6 sp1 -mask -p 4 -bits
```

#### Output

A folder with motif information(.html files).
Then we can extract Position Weight Matrix(pwm) from these html files and visualize motifs.

### Citations

> Heinz S, Benner C, Spann N, Bertolino E et al. Simple Combinations of Lineage-Determining Transcription Factors Prime cis-Regulatory Elements Required for Macrophage and B Cell Identities. Mol Cell 2010 May 28;38(4):576-589. PMID: 20513432
